internal/dispatch/caching/caching.go

```diff
@@ -183,7 +188,24 @@ func (cd *Dispatcher) SetDelegate(delegate dispatch.Dispatcher) {
 // DispatchCheck implements dispatch.Check interface
 func (cd *Dispatcher) DispatchCheck(ctx context.Context, req *v1.DispatchCheckRequest) (*v1.DispatchCheckResponse, error) {
 	cd.checkTotalCounter.Inc()
+
+	// Load the relation to get its computed cache key, if any.
+	// NOTE: We do not use the canonicalized cache key when checking within the same namespace, as
+	// we may get different results if the subject being checked matches the resource exactly.
 	requestKey := dispatch.CheckRequestToKey(req)
+	if cd.nsm != nil && req.ObjectAndRelation.Namespace != req.Subject.Namespace {
+		revision, err := decimal.NewFromString(req.Metadata.AtRevision)
+		if err != nil {
+			return nil, err
+		}
+
+		_, relation, err := cd.nsm.ReadNamespaceAndRelation(ctx, req.ObjectAndRelation.Namespace, req.ObjectAndRelation.Relation, revision)
+		if err != nil {
+			return nil, err
+		}
+
+		requestKey = dispatch.CheckRequestToKeyWithPossibleCanonical(req, relation.CanonicalCacheKey)
```

Comment:
If there is no canonical cache key, this method calls requestKey to re-compute the key for the check request. Seems wasteful to do it twice...